src/main/java/com/team766/controllers/ExtendableCanSparkMaxSmartMotionPIDController.java

```diff
@@ -0,0 +1,283 @@
+package com.team766.controllers;
+
+import com.team766.config.ConfigFileReader;
+import com.team766.hal.RobotProvider;
+import com.team766.library.SetValueProvider;
+import com.team766.library.SettableValueProvider;
+import com.team766.library.ValueProvider;
+import com.team766.logging.Logger;
+import com.team766.logging.LoggerExceptionUtils;
+import edu.wpi.first.math.MathUtil;
+import com.revrobotics.SparkMaxAbsoluteEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.ControlType;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.SparkMaxAbsoluteEncoder.Type;
+import java.io.IOError;
+import javax.swing.text.DefaultStyledDocument.ElementSpec;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.revrobotics.CANSparkMax;
+import com.team766.framework.Mechanism;
+import com.team766.hal.MotorController;
+import com.team766.library.RateLimiter;
+import com.team766.controllers.pidstate.*;
+import com.team766.framework.Exceptions.*;
+
+
+public abstract class ExtendableCanSparkMaxSmartMotionPIDController {
+		/*
+		 * These are the hardware variables for the hardware we have on the robot.
+		 * It includes a CanSparkMax, MotorController of any type, and a SparkMaxAbsoluteEncoder.
```

Comment:
I thought they were? Motorcontroller is like the rev neo and CSM is the motorcontroller. This is bad naming. We should fix it.

Reply:
They are the same device. They are even the same Java object - it is a [`CANSparkMaxMotorController`](https://github.com/Team766/MaroonFramework/blob/master/src/main/java/com/team766/hal/wpilib/CANSparkMaxMotorController.java), which extends `CanSparkMax` and implements `MotorController`, so it can be cast to either of those types

Reply:
i.e. why you can do `csm = (CANSparkMax) mc;` in the constructor

Reply:
Interesting. From past testing, this is what works with Maroon Framework. We could try it the other way, but I'm afraid that removing the ```MotorController``` declaration won't allow us to set power to the motors correctly.

Reply:
Ryan's right; you can actually do away with `mc` as `csm` points to the same object, just specifically casted to expose extra `CanSparkMax` methods/properties.

Reply:
whatever methods and properties `MotorController` has, `CANSparkMax` would too.

Reply:
I get this - but I think that the way we use the motor stuff from the config file makes this work better.